Fixes some player's origin displacements for fire events, impulses commands…

XutaxKamay opened the pull request:
This fixes shooting problems and others things, for example when you're running and aim perfectly at the head of someone.

https://www.youtube.com/watch?v=VPT0-CKODNc

Why is this happening?

Well, this is happening mainly because of interpolation. Let me explain.
In fact, the taken shooting position is the new one computed from the game's movement, and not the current one. You can see this happening by using timescale, and getting some speed. The game is computing the new interpolated localplayer's data (@BaseInteporlate1 from C_BaseEntity)  by getting the final predicted time (so the final predicted command) and removes a tick interval to actually take the previous final predicted time ,  _**for having data to interpolate with!**_ , and adds a tick interval where it's mulitplied by interpolation_amount (this is computed per frames, and it goes to 0.0 to 1.0) to go on the final predicted time, wich finally goes to what you should see. If you try to interpolate directly from the final predicted time, there wouldn't be any data to interpolate with, so that's why valve is removing a tick interval on the final predicted time to interpolate the localplayer.
But it makes issues on player's gameplay we must remember that since the commands are ran per ticks, the changes are instantaneous, so when you're actually firing, you're still on the previous predicted time, and not the final one.
 You see the world as it was previously and not as it was when firing.
The solution of what you would think of would be of course to disable interpolation on a player side, but that would not make a smooth gameplay (so makes still a game unplayable) and player couldn't even predict his next position to actually shoot at the head of the player correctly anyway because changes on player's will be instantaneous. (or he must be a real machine to do that). Because since the game is running the commands per ticks , you'll suddently "teleport" to the other side and see your bullet at that position, instead of seeing your shot behind, as expected and predicted by the player. So the solution is just to call PostThink, before applying the final data to the player, so interpolation is not a problem anymore, the weapons takes the right origin to shoot with, and much others stuffs! Like collision bounds, or when you're actually landing with high velocities, it stops the falling velocity at where it should and not 1 tick before. (bunny hopping purposes)

https://steamuserimages-a.akamaihd.net/ugc/909030263364841206/048560EB944800B68FF1F52920763D5B66231AE7/

Here is a view example of what is happening, blue = what the game actually see, red = what the player saw in the game.

I hope valve and people will understand what I explained and hopefully this gets fixed on all valve games so we get a proper gameplay for everyone! (Yes since I guess valve is using this code for every games, it needs to be fixed on every games)

Good day and thanks for atleast reading!

PS: Sorry for my english if it's not correct, but if I get understood, it's the more important.

TotallyMehis commented:
Here's a test I did:
https://youtu.be/VYs0-lUwWas

Seems to hold up.

Sh1ft0x0EF commented:
There are many videos on YouTube about Source hitreg in such games as CS:GO and especially TF2 which clearly shows a direct headshot that wasn't registered as a hit by the engine
Source hitreg is one of those "WTF" things about the engine
The fun part is that Source contains a lot of smoothing and compensation mechanisms but GS simply works better without all of them

Regarding your explanation - there should be interpolation for remote (server-side) entities on client but the thing is...

>But it makes issues on player's gameplay we must remember that since the commands are ran per ticks, the changes are instantaneous, so when you're actually firing, you're still on the previous predicted time, and not the final one.
You see the world as it was previously and not as it was when firing.
The solution of what you would think of would be of course to disable interpolation on a player side, but that would not make a smooth gameplay (so makes still a game unplayable) and the player wouldn't even predict his next position to actually shoot at the head of the player correctly anyway (or he must be a real machine to do that) , but why?

Isn't this issue should be handled by the lag-compensation mechanism?

Well, anyway, if your PR is actually fixing the "broken" hitreg of Source games then personally for me this would be pretty valuable contribution to the codebase

XutaxKamay commented:
Well in case you got confused, it's not about others entities, but the local player, so it's a totally different thing, and interpolation is working pretty much different here, as I explained. The interpolation for others entities where they aren't local entities (local player), use the commands based on cl_interp , cl_interp_ratio, cl_updaterate and many others for interpolating. But local player, use the prediction data for interpolating, wich is why it would be absurd to use it with lag compensation. Lag compensation must NOT be active on local player, even if it could be used for fixing here , because the prediction is already (atleast should be) compensating for it , so there is no point to use lag compensation on a local entity. If you try to make lag compensation on a local player, the client side data wouldn't even correspond to server's and make prediction errors, wich makes finally non sences to use it on a local player, since prediction is here for it anyway as I said earlier. (CS:GO has servers with a custom random seed, but that's another story, it was meant to stop cheaters abusing the prediction for their aimbots and etc..). But like you said, interpolation is taken into account for others entities in lag compensation, but definitly not for a local player. Also this fix every valve games (not source engine only), because even CS:GO or even new games from valve should have that bug, since I don't think they changed server runcommand and client prediction since then. The bug exists since the first very engine of valve is born I guess. I just hope that valve will look at it and make a proper fix for it. On a client side (so local), you could always "over predict" your commands, so you can be at where you should be shooting at, but it's not very accurate depending what movements you're doing etc. I hope I replied to your question. Thanks for reading.

notsenator commented:
You also have to remember that because the server beholds a random seed that’s purposely different to the clients to stop the client predicting the seed, the shot you fire doesn’t appear where you see it on your screen because the seeds are different and not given to the player to prevent calculation of the spread thus nospread

Sh1ft0x0EF commented:
@XutaxKamay 
There is no point in running lag compensation on client-side
Client should just set its cvar setting to true for lag comp and that cvar will be sent to server where, according to the value of that cvar, it will treat the client specially (both GS and Src works that way with minor diffs)

Did you read that?:
https://developer.valvesoftware.com/wiki/Source_Multiplayer_Networking#Lag_compensation
https://developer.valvesoftware.com/wiki/Lag_compensation

> I just hope that valve will look at it and make a proper fix for it, since it makes the game much more accurate

Well, they recently updated GoldSrc (after 4 years) with some bug fixes that were on halflife repo issue tracker, so yeah, maybe they will... in 2020

XutaxKamay commented:
Exactly there is no point running client side compesation i didnt mean something else.

geotavros commented:
Are you sure this issue is not related to the slow motion you enable to test it? I mean, did anyone test it while not in slow motion? For example, recording a 300 fps video while the game is in real-time and then checking this video out.

XutaxKamay commented:
Though this pullrequest break with physics props still, so I'm going probably to commit another fix for it. The simple solution would be to take the old origin of the player before calling the game movement and set this origin for shooting position , but it won't fix others things.

XutaxKamay commented:
2 years later.. I really need to re-work on it properly.. That'll come.

Kefta commented:
We are eagerly awaiting

Everlastingbulletlag commented:
Why is this NOT solved yet Volvo? Are all their developers nobbs or wtf is going on. I have got this for 17 years and the idiots still havent fix it. I dont play their games anymore and I dont buy new before they wake up. Its 2020 now and they havent admitted yet that they fucked up. The developers respsible for this should get lifesensen in jail. It has made people sick and destroying their life now I hope the people responsible wil get their punishment. Its serious the worse company ever. The change this and fucked up and none of them has guts to admit and fix it. Its like they are afffaid of what will happends when people find out who responsible. I understand if people will find them and destroy their lifes as they destroyed ours.

XutaxKamay commented:
Pls no hate, love.
I'm sure they're aware of it but probably busy with some others stuffs we don't know yet? I know your frustration though I wish I could at least make signed Valve Client Plugins so I can load custom stuff inside game's client to fix some bugs or add functionalities, I've even tried to email them about it but I did not get any replies yet.
Anyway yes you made me remind that I need to re-work on it haha

Kefta commented:
We'll be here when you finish it king.

karmakarmakarmakarmakarmakarmakarma commented:
thanks for approving this KillaBoi

KillaBoi commented:
> thanks for approving this KillaBoi

I did this like a year ago, Valve doesn't care about it :(

XutaxKamay commented:
Hey it's been 2 years again, I said I should have worked on it, but guess what...
I did nothing. ( ͡° ͜ʖ ͡°)

Time passes, and now I realize it's been already 4 years I did that pullrequest and it has been 10 years since I've been starting to code on that engine.

No really, I think this should be on top of my priorities so it can be even considered to be implemented in valve's code, like I said they probably waiting for me in secret to finish it... I must tell that they're not very communicative sometimes, probably busy I don't know.

I saw that there was a CS:GO update about making custom client plugins that could be used to fix the game, I think I tried to email them about it so mine could get signed by them, but I don't even know if they saw it.
I mean I could be even ready to share it publicly with source code available with some quality fixes/changes.

KillaBoi commented:
> Hey it's been 2 years again, I said I should have worked on it, but guess what...
> I did nothing. ( ͡° ͜ʖ ͡°)
> 
> Time passes, and now I realize it's been already 4 years I did that pullrequest and it has been 10 years since I've been starting to code on that engine.
> 
> No really, I think this should be on top of my priorities so it can be even considered to be implemented in valve's code, like I said they probably waiting for me in secret to finish it... I must tell that they're not very communicative sometimes, probably busy I don't know.
> 
> I saw that there was a CS:GO update about making custom client plugins that could be used to fix the game, I think I tried to email them about it so mine could get signed by them, but I don't even know if they saw it.
> I mean I could be even ready to share it publicly with source code available with some quality fixes/changes.

sharklaser tried to fix desync/antiaim back in 2018 and even released source code to fix it and emailed them about it but they didn't even bother with fixing that so i doubt they'll implement any of this tbh, it sucks but theres only so much u can do... you can release it if u think it'll help modders using the sdk in the future or smth but ye :/

XutaxKamay commented:
Well, if its purpose is to be used on CS:GO it is quite useless because matchmaking servers are hosted by Valve. So I doubt to see it any of it soon. 
But the community servers can already have that fix for their shooting position I've made one for sourcemod for almost all games.

osiris-plus commented:
Will never be fixed.

micwoj92 commented:
> Will never be fixed.

Who are you to say that, Gaben Newell?

KillaBoi commented:
> > Will never be fixed.
> 
> Who are you to say that, Gaben Newell?

Gaben Newell OMEGALUL, no it's Gabe Logan Newell and it's quite obvious that most things in this repo aren't going to be implemented by Valve, all these PR's are there for future devs to potentially add into their projects

micwoj92 commented:
> Gaben Newell OMEGALUL, no it's Gabe Logan Newell and it's quite obvious that most things in this repo aren't going to be implemented by Valve, all these PR's are there for future devs to potentially add into their projects

It's Gaben Newell.

Kxnrl commented:
> > Hey it's been 2 years again, I said I should have worked on it, but guess what...
> > I did nothing. ( ͡° ͜ʖ ͡°)
> > Time passes, and now I realize it's been already 4 years I did that pullrequest and it has been 10 years since I've been starting to code on that engine.
> > No really, I think this should be on top of my priorities so it can be even considered to be implemented in valve's code, like I said they probably waiting for me in secret to finish it... I must tell that they're not very communicative sometimes, probably busy I don't know.
> > I saw that there was a CS:GO update about making custom client plugins that could be used to fix the game, I think I tried to email them about it so mine could get signed by them, but I don't even know if they saw it.
> > I mean I could be even ready to share it publicly with source code available with some quality fixes/changes.
> 
> sharklaser tried to fix desync/antiaim back in 2018 and even released source code to fix it and emailed them about it but they didn't even bother with fixing that so i doubt they'll implement any of this tbh, it sucks but theres only so much u can do... you can release it if u think it'll help modders using the sdk in the future or smth but ye :/

I can not find that pr or src from github/google, can you share the link?

7kazan commented:
How can I apply this fix to no more room in hell ? I tried here it doesn't work

XutaxKamay commented:
Fun fact; if you do cl_predict 0, it still demonstrates perfectly the problem described in the pull request.

Like I said, this isn't a problem with interpolation much in the first place. (I haven't reverse engineered the game yet, I just know they take the eye position from here, that command prooves it, which is an issue)

It also seems they fixed prediction for their sub-ticks too, which is a good thing, it was really an issue before.

It's gonna take a while for them too to implement it properly by applying a fraction (gpGlobals->interpolation_amount or something else) everywhere in the game gamemovement and physics, Think functions.
Sadly even with those fixes etc, but it will stay mostly with the same inconsitencies simply because the game isn't based anymore on constant tickrates.
Performance wise, it's not really great either since it's dommed to run more ticks than constant tickrates.

I won't bash at Valve I think they expected us to be happy with it but to be honest I would be happier with them removing it for now and thinking of another solution to fix interpolation problems that doesn't affect tickrate.

Nairdaa commented:
I'm so glad Kamay showed up again and is back with great input. I honestly hope Valve does take a deeper look into what they have done (and in my opinion, broke) and fix the game.

XutaxKamay commented:
Found another way to "fix" it by looking into engine code a bit more.
The indentation is broken, I know:

https://github.com/XutaxKamay/css_enhanced_waf/commit/fbc45f4a64173d4d64aed01ed9f1e99a44dcb744

Right now like I said in the commit with high enough fps it is not noticeable and does not require any of the above solution(s) I mentioned, the downside is that it might be a bit jittering when you don't have enough fps, but otherwise it runs pretty smoothly and does not require any user commands modification and fixes some local player interpolation issues. (finally, the user commands matching the screen)

That still requires though fixing the shooting position being 1 tick ahead.

![image](https://github.com/user-attachments/assets/aa642430-e368-4bc5-a309-580f0c8a9512)

&

https://xutaxkamay.com/videos/fix.avi

For a bit of context this screenshot was taken when FireBullet is fired using noclip without timescale, notice the velocity and precision here. (ignore punchangles)

What I can say now though, is that objectively sub-tick (frame based buttons) has still its uses for having inputs buttons less delayed and gain some advantage of not wanting a tick to happen, but I'll keep saying that it shouldn't be used for movements at all.

XutaxKamay commented:
Turns out it was very easy to smooth out aswell:

https://github.com/XutaxKamay/css_enhanced_waf/blob/519afd312fb1b93a369f23e5056789249622ec8f/engine/host.cpp#L3189

Just substracting the old interpolation amount that wasn't taken into account will do the trick

For reference, here is 1 frame, 1 tick, 9 frames, 1 tick sequence:
```
current interp: 0.954998, old amount: 0.987125, time: 0.003507, frametime: 0.003507, last remainder not interpolated: 0.032126
interpolation amount was 0.039147, corrected to fix interpolation issues.
current interp: 0.146476, old amount: 0.185623, time: 0.004577, frametime: 0.004577, last remainder not interpolated: 0.039147
current interp: 0.241666, old amount: 0.280812, time: 0.002975, frametime: 0.002975, last remainder not interpolated: 0.039147
current interp: 0.340323, old amount: 0.379470, time: 0.003083, frametime: 0.003083, last remainder not interpolated: 0.039147
current interp: 0.438092, old amount: 0.477238, time: 0.003055, frametime: 0.003055, last remainder not interpolated: 0.039147
current interp: 0.567577, old amount: 0.606724, time: 0.004046, frametime: 0.004046, last remainder not interpolated: 0.039147
current interp: 0.649311, old amount: 0.688457, time: 0.002554, frametime: 0.002554, last remainder not interpolated: 0.039147
current interp: 0.787602, old amount: 0.826749, time: 0.004322, frametime: 0.004322, last remainder not interpolated: 0.039147
current interp: 0.854431, old amount: 0.893578, time: 0.002088, frametime: 0.002088, last remainder not interpolated: 0.039147
current interp: 0.930525, old amount: 0.969672, time: 0.002378, frametime: 0.002378, last remainder not interpolated: 0.039147
interpolation amount was 0.050799, corrected to fix interpolation issues.

```

As you can see, 0.930525 is a lot better than 0.969672 on the last frame to go on a new tick with 0 interpolation amount again (so going the new position).
On the first frame too, after the new tick arrival, going from 0.0 to 0.146476 instead of 0.185623 is a lot better for smooth interpolation too!

32 tickrate, 33 fps:

```
interpolation amount was 0.003660, corrected to fix interpolation issues.
current interp: 0.969697, old amount: 0.973357, time: 0.030303, frametime: 0.030303, last remainder not interpolated: 0.003660
interpolation amount was 0.943054, corrected to fix interpolation issues.
interpolation amount was 0.912752, corrected to fix interpolation issues.
interpolation amount was 0.882449, corrected to fix interpolation issues.
interpolation amount was 0.852146, corrected to fix interpolation issues.
interpolation amount was 0.821844, corrected to fix interpolation issues.
interpolation amount was 0.791541, corrected to fix interpolation issues.
interpolation amount was 0.761239, corrected to fix interpolation issues.
interpolation amount was 0.730936, corrected to fix interpolation issues.
interpolation amount was 0.700634, corrected to fix interpolation issues.
interpolation amount was 0.670331, corrected to fix interpolation issues.
interpolation amount was 0.640029, corrected to fix interpolation issues.
interpolation amount was 0.609727, corrected to fix interpolation issues.
interpolation amount was 0.579424, corrected to fix interpolation issues.
interpolation amount was 0.549121, corrected to fix interpolation issues.
interpolation amount was 0.518819, corrected to fix interpolation issues.
interpolation amount was 0.488516, corrected to fix interpolation issues.
interpolation amount was 0.458214, corrected to fix interpolation issues.
interpolation amount was 0.427912, corrected to fix interpolation issues.
interpolation amount was 0.397610, corrected to fix interpolation issues.
interpolation amount was 0.367307, corrected to fix interpolation issues.
interpolation amount was 0.337005, corrected to fix interpolation issues.
interpolation amount was 0.306703, corrected to fix interpolation issues.
interpolation amount was 0.276400, corrected to fix interpolation issues.
interpolation amount was 0.246097, corrected to fix interpolation issues.
interpolation amount was 0.215795, corrected to fix interpolation issues.
interpolation amount was 0.185492, corrected to fix interpolation issues.
interpolation amount was 0.155190, corrected to fix interpolation issues.
interpolation amount was 0.124887, corrected to fix interpolation issues.
interpolation amount was 0.094592, corrected to fix interpolation issues.
interpolation amount was 0.064289, corrected to fix interpolation issues.
interpolation amount was 0.033987, corrected to fix interpolation issues.
interpolation amount was 0.003684, corrected to fix interpolation issues.
current interp: 0.969698, old amount: 0.973381, time: 0.030303, frametime: 0.030303, last remainder not interpolated: 0.003684
```

Seems like good enough to me.

XutaxKamay commented:
Tried to make it a bit smoother, https://github.com/XutaxKamay/css_enhanced_waf/blob/master/engine/host.cpp#L3194
I'd like to keep in mind that this isn't a very perfect method,  to be honest it might be debatable if it's even worth it since it might be a less better in terms of smoothness, but it's *very* subtle now.

XutaxKamay commented:
Don't mind the code, it's not the cleanest.

https://github.com/XutaxKamay/css_enhanced_waf/blob/956d26bcb303cc03e34886216be073ad6afb15c0/game/shared/cstrike/fx_cs_shared.cpp#L155

This is the method by using interpolation_amount, I finally took this method because it made the game a bit unsmooth previously even if it was very subtle, it also has been noticed by some others playtesters.

By doing this, I have noticed a problem also with interpolation_amount, the fact that it gets calculated after CreateMove makes it late by one frame. I believe this problem is also on CS2.

https://github.com/XutaxKamay/css_enhanced_waf/blob/956d26bcb303cc03e34886216be073ad6afb15c0/engine/host.cpp#L2659

The variable needs to be recalculated before CL_Move is called and restore it to previous value to avoid any interpolation problems. Some people might prefer to take the shoot position the last frame before a new tick arrival though, but if you want to match the interpolated position on the same exact tick it fires, this is what you want.

darthelmo1 commented:
Is it finally time? @XutaxKamay PRs are getting accepted...

XutaxKamay commented:
> Is it finally time? @XutaxKamay PRs are getting accepted...

Doubt it's going to get accepted anyway but I realized that this needs some engine modifications to do it properly, so this "can't be done" (you can but you need to hook functions inside the engine, PITA)
